ai_genomics/utils/openalex.py

```diff
@@ -0,0 +1,144 @@
+
+from ai_genomics import get_yaml_config, _base_config_path
+
+CONCEPT_THRES = get_yaml_config(_base_config_path)["concept_threshold"]
+
+
+def deinvert_abstract(inverted_abstract: Dict[str, List]) -> str:
+    """Convert inverted abstract into normal abstract
+
+    Args:
+        inverted_abstract: a dict where the keys are words
+        and the values lists of positions
+
+    Returns:
+        A str that reconstitutes the abstrat
+
+    """
+
+    abstr_empty = (max(chain(*inverted_abstract.values())) + 1) * [None]
+
+    for word, pos in inverted_abstract.items():
+        for p in pos:
+            abstr_empty[p] = word
+
+    return " ".join(abstr_empty)
+
+
+def get_institutions(doc: dict) -> list:
+    """
+    Create a list of all institutions involved in a work
+    """
+
+    return [
+        inst["id"]
+        for auth in doc["authorships"]
+        for inst in auth["institutions"]
+        if len(inst) > 0
+    ]
+
+
+def enrich_institutions(doc_inst: list, insts_dict: dict, variable: str):
+    """
+    Create institutional variables for a list of institutions
+    involved in a work
+    """
+
+    return [insts_dict[inst][variable] if inst != None else None for inst in doc_inst]
+
+
+def get_concepts(doc: dict, thres: float = CONCEPT_THRES) -> list:
+    """Create a list of concepts associated with a work
+
+    Args:
+        doc: OA work
+        thres: minimum threshold for inclusion
+
+    """
+
+    return [
+        concept["display_name"]
+        for concept in doc["concepts"]
+        if float(concept["score"]) > thres
+    ]
+
+
+def make_inst_lookup(inst_dict: dict, variable: str):
+    """Create a lookup between inst id and variable
+
+    Args:
+        inst_dict: a lookup between inst ids and their metadata
+        variable: the variable to look up
+
+    """
+    return {key: value[variable] for key, value in inst_dict.items()}
+
+
+def work_metadata_df(
+    work_list: list,
+    inst_list: list,
+    meta_vars: list = ["country_code", "type", "display_name"],
+) -> DataFrame:
+    """
+    Creates a dataframe with work metadata
+
+    Args:
+        work_list: a list of OA works
+        inst_list: a list of OA institutions
+        meta_vars: list of metadata variables to include in the dataframe
+
+    Returns:
+        A dataframe where rows are works and columns are metadata
+        of interest
+    """
+
+    inst_dict = {inst["id"]: inst for inst in inst_list}
+
+    inst_df = (
+        pipe(
+            [
+                (doc["id"], doc["publication_year"], get_institutions(doc))
+                for doc in work_list
+            ],
+            partial(DataFrame, columns=["work_id", "year", "institution"]),
+        )
+        .explode("institution")
+        .reset_index(drop=True)
+    )
+
+    # We create a column for every new institutional metadata variable
+    for var in meta_vars:
+        inst_df[var] = inst_df["institution"].map(make_inst_lookup(inst_dict, var))
```

Comment:
Not really a big performance issue in this instance, but this method means that you have to cycle through all institutions to create a new lookup for each meta variable. For bigger data/more iterations if it might be faster to create a new data frame with all the meta variables and merge or to update a list of inst dicts with the meta variable dicts and then construct a dataframe at the end.